lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,361 @@
+        txWithCerts [] = Nothing
+        txWithCerts cs = Just $ Tx
+            { certs = cs
+            , inputs = maybeToList (mkTxIn <$> pointer s)
+            , outputs =
+                [ TxOut
+                    (mkAddress $ keyAtIx s $ nextKeyIx s')
+                    (TB.fromCoin minUTxOVal)
+                ]
+            }
+    in
+        txWithCerts $ reRegKey0IfNeeded
+            ++ deRegKey0IfNeeded
+            ++ case compare (toEnum n) (nextKeyIx s) of
+                GT -> deleg [nextKeyIx s .. toEnum (n - 1)]
+                EQ -> []
+                LT -> dereg $ reverse [toEnum n .. (pred $ nextKeyIx s)]
+  where
+    deleg = (>>= \ix ->
+        if isReg (acct ix)
+        then [Delegate (acct ix)]
+        else [RegisterKey (acct ix),  Delegate (acct ix)]
+        )
+
+    reRegKey0IfNeeded =
+        if not (isKey0Reg s) && n > 0 && fromEnum (nextKeyIx s) > 1
+        then [RegisterKey (acct minBound), Delegate (acct minBound)]
+        else []
+
+    deRegKey0IfNeeded =
+        [ DeRegisterKey (acct minBound)
+        | (isKey0Reg s)
+        , n == 0
+        , fromEnum (nextKeyIx s) == 0
+        ]
+
+    dereg ixs =
+        [ DeRegisterKey (acct ix)
+        | ix <- ixs
+        , isReg . acct $ ix
+        -- We need to /at least/ check @isReg (key 0)@, because the user could
+        -- have deregistered it using old wallet software.
+        ]
+
+    acct = toRewardAccount . keyAtIx s
+
+applyTx
+    :: forall k. ( SoftDerivation k
+        , ToRewardAccount k
+        , MkKeyFingerprint k Address
+        , MkKeyFingerprint k (k 'AddressK XPub))
+    => Tx
+    -> Hash "Tx"
+    -> DelegationState k
+    -> DelegationState k
+applyTx (Tx cs _ins outs) h s0 =
+    let
+        s = foldl (flip applyCert) s0 cs
+        isOurOut i (TxOut addr _b) =
+            case (paymentKeyFingerprint @k $ keyAtIx s i, paymentKeyFingerprint addr) of
+            (Right fp, Right fp')
+                | fp == fp' -> True
+                | otherwise -> False
+            _ -> False
+        -- To workaround rollback + old wallets, we also look for the pointer at
+        -- 0 and 1, aside from nextKeyIx.
+        f o = any (`isOurOut` o) [ nextKeyIx s, minBound, succ minBound ]
+        pointerOuts = filter (f . snd) $ zip [0..] outs
+    in case pointerOuts of
+        [] -> s --error "panic: no pointer utxo on-chain" -- What should we do?
+        [(ix,TxOut _addr tb)]
+            -> s { pointer = Just $ PointerUTxO (TxIn h ix) (TB.getCoin tb) }
+        (_x:_) -> error "todo"
+            -- There shouldn't be more than one pointer, but theoretically
+            -- possible. If a user sends funds to an address corresponding to
+            -- the stake key (why would they do this though?), we could mistake
+            -- it for the pointer.
+            --
+            -- TODO: Is this actually bad though? Or is it good because it will
+            -- then be sent back as change? Any other problems?
```

Comment:
> by requiring that the pointer can only be a UTxO that is part of a transaction which also contains delegation certificate or key de-registration certificates

Now done.

> It however forces the wallet to "follow the pointer" in a way, and simply querying the LSQ isn't sufficient to figure out whether a pointer exist. I think that following the pointer was however already necessary in order to spend it right (we need to have the exact TxIn for that) ?

I haven't put much thought about using LSQ here, but I don't see why it would help. Now we do need to store a `TxIn`, but it's not that complicated.